akka-cluster-typed/src/main/scala/akka/cluster/typed/internal/receptionist/ClusterReceptionist.scala

```diff
@@ -116,7 +150,42 @@ private[typed] object ClusterReceptionist extends ReceptionistBehaviorProvider {
           externalInterface.RegistrationsChangedExternally(changes, newState)
       }
 
-    replicator ! Replicator.Subscribe(ReceptionistKey, adapter.toUntyped)
+    replicator ! Replicator.Subscribe(ReceptionistKey, replicatorMessageAdapter.toUntyped)
+
+    // remove entries when members are removed
+    val clusterEventMessageAdapter: ActorRef[MemberRemoved] =
+      ctx.messageAdapter[MemberRemoved] {
+        case MemberRemoved(member, _) ⇒
+          // ok to update from several nodes but more efficient to try to do it from one node
+          if (cluster.state.leader.contains(cluster.selfAddress)) {
+            if (member.address == cluster.selfAddress) NodesRemoved.empty
+            else NodesRemoved(Set(member.address))
+          } else
+            NodesRemoved.empty
+      }
+
+    cluster.subscribe(clusterEventMessageAdapter.toUntyped, ClusterEvent.InitialStateAsEvents, classOf[MemberRemoved])
+
+    // also periodic cleanup in case removal from ORMultiMap is skipped due to concurrent update,
```

Comment:
In which cases is the removal lost? Can it happen for any removal of the same key? Does the same problem then apply to regular unregistering as well? Can it happen that adding an entry for a key on another node will lead to a missed removal?

What will happen if a service registers and unregisters itself periodical quickly? (I guess it might stay registered even if the last operation was unregister?) What will happen if the nodes goes up and down quickly? (Is it the "node registers / node fails" before ddata has converged scenario that is tried to be fixed by the below code?)

Reply:
> Does the same problem then apply to regular unregistering as well?

You mean if a registered actor is terminated, but the node is still up and running.
yes, I think it's the semantics of the ORSet, so that could also be a problem.
I'll keep the ticket open and add this as another thing to consider. That is even more difficult to solve